spring-core/src/main/java/org/springframework/util/ResourceUtils.java

```diff
@@ -384,4 +389,23 @@ public static void useCachesIfNecessary(URLConnection con) {
 		con.setUseCaches(con.getClass().getSimpleName().startsWith("JNLP"));
 	}
 
+	/**
+	 * Determine, if possible, the contentLength of the given resource
+	 * without reading it.
+	 * @param resource the resource instance
+	 * @return the contentLength of the resource
+	 */
+	public static OptionalLong contentLength(Resource resource) {
+		// Don't try to determine contentLength on InputStreamResource - cannot be read afterwards...
+		// Note: custom InputStreamResource subclasses could provide a pre-calculated content length!
+		if (InputStreamResource.class != resource.getClass()) {
+			try {
+				return OptionalLong.of(resource.contentLength());
+			}
+			catch (IOException ignored) {
+			}
+		}
+		return OptionalLong.empty();
+	}
+
```

Comment:
Could we consider a new method on Resource for this? The fact that a utility method takes a Resource and calls its property is a sign. It looks like a variant of the current `getContentLength()` that returns the content length optionally if known. So perhaps `OptoinalLong getContentLengthIfKnown()` with a default implementation returning empty.

/cc @jhoeller, @poutsma